### PR TITLE
Capital letters

### LINKED ISSUE #14: Challenge_Convert_Letters_Capital
Writing a function that converts letters to capital 

input=(Hello)
output=(HELLO)

### PR DESCRIPTION
**Pull request by Kareiman**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_convert_to_capital.py
Comment (Frank2446-dotcom) [inline solutions/tests/test_convert_to_capital.py:1]: The test file demonstrates thoughtful and thorough coverage of different cases, including standard, edge, and defensive tests. It ensures the function works as intended, even with unconventional inputs like special characters or mixed content. Great attention to detail in testing all possible edge cases—well-done!

File: solutions/convert_to_capital.py
Comment (Frank2446-dotcom) [inline solutions/convert_to_capital.py:1]: The implementation is well-structured and effectively handles edge cases, such as empty inputs and mixed character types. Excellent use of input validation and logical organization, ensuring the function performs reliably for various scenarios. Well-done job on ensuring clarity and functionality!

**Review by Frank2446-dotcom — APPROVED** [review]
The challenge is tackled effectively, with a well-implemented solution and comprehensive testing to ensure functionality across a variety of scenarios, including standard, edge, and defensive cases. Excellent work on maintaining clarity, input validation, and robust test coverage, demonstrating a thorough understanding of the task—well done!

File: solutions/tests/test_convert_to_capital.py
Comment (Kareiman) [reply solutions/tests/test_convert_to_capital.py:1]: Thank you @Frank2446-dotcom for the kind words and the thoughtful feedback.
 I appreciate you noticing the details—it means a lot!

File: solutions/convert_to_capital.py
Comment (Kareiman) [reply solutions/convert_to_capital.py:1]: Thank you for your detailed feedback!

File: solutions/convert_to_capital.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# !/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+"""
+A module that converts letters to upper case
```
Comment (Emanfalouji) [inline solutions/convert_to_capital.py]: just correct that uppercase is a one word

File: solutions/convert_to_capital.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+@author: Kareiman Altayeb
+"""
+
+
+def convert_to_capital(user_text: str) -> str:
+    """Asks the user to enter a text and returns the text in capital
+
+    parameters:
+        user_text = in str
```
Comment (Emanfalouji) [inline solutions/convert_to_capital.py]: it could be corrected to this format 
user_text (type): description      
like
 user_text (str): The user input text to be converted to uppercase.

File: solutions/convert_to_capital.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    returns:
+        user_text in capital letters
```
Comment (Emanfalouji) [inline solutions/convert_to_capital.py]: you could add the return type here to be like this 
(str) : user_text in capital letters

File: solutions/convert_to_capital.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    raises:
+        AssertionError: if input is empty
```
Comment (Emanfalouji) [inline solutions/convert_to_capital.py]: could you add as you mentioned in the assrtionError message to be like this 
 AssertionError: If the input is empty or contains only spaces.

File: solutions/convert_to_capital.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+    examples:
```
Comment (Emanfalouji) [inline solutions/convert_to_capital.py]: could you rewrite Examples with capital E  
also the previous lines , to start with capital letter (Raises , Returns and Parameters)

File: solutions/tests/test_convert_to_capital.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
+# !/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+"""
+A module to test convert_to_capital function
+
```
Comment (Emanfalouji) [inline solutions/tests/test_convert_to_capital.py:6]: could you add here the Test categories with Standard cases, Edge cases, and Defensive tests

**Review by Emanfalouji — CHANGES_REQUESTED** [review]

File: solutions/tests/test_convert_to_capital.py (hunk 6 above)
Comment (Kareiman) [reply solutions/tests/test_convert_to_capital.py:6]: done

File: solutions/convert_to_capital.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+This function asks user to enter a word and converts the letters
+to capital. The function will return numbers and capital letters the same
+
+Created on 31 12 2024
+
+@author: Kareiman Altayeb
+"""
+
+
+def convert_to_capital(user_text: str) -> str:
```
Comment (Frank2446-dotcom) [inline solutions/convert_to_capital.py:13]: 1. Current: convert_to_capital

Improvement: While this name is clear, it could be more explicit in describing what the function does, such as convert_to_uppercase, to clearly convey that it deals with uppercase conversion. "Capital" can be ambiguous because people might think it deals with capitalization rules (e.g., first letter capitalized), not case swapping.

Suggested Improvement: convert_to_uppercase

File: solutions/convert_to_capital.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+# -*- coding: utf-8 -*-
+
+"""
+This function asks user to enter a word and converts the letters
+to capital. The function will return numbers and capital letters the same
+
+Created on 31 12 2024
+
+@author: Kareiman Altayeb
+"""
```
Comment (Frank2446-dotcom) [inline solutions/convert_to_capital.py:10]: Improvement: The module docstring does not specify the function’s behavior as clearly as it could. Consider improving the explanation of the function and its purpose.

"""
This function takes a user input string and returns the string with all characters in uppercase. 
It handles edge cases where the input is empty or contains only spaces.
"""

File: solutions/convert_to_capital.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+        'HELLO'
+    >>> convert_to_capital('123hello')
+        '123HELLO'
+    """
+
+    user_text = user_text.strip()
+
+    if not user_text:
+        raise AssertionError("Entry cannot be empty or just spaces")
+
```
Comment (Frank2446-dotcom) [inline solutions/convert_to_capital.py:38]: Improvement: The use of strip() and then checking if not user_text is good. However, you can directly handle empty strings in the assertion, which makes the function simpler:
Action: Use strip() to handle spaces at the beginning or end of the input more cleanly in the assertion check.

e.g

if not user_text.strip():
    raise AssertionError("Entry cannot be empty or just spaces")

File: solutions/tests/test_convert_to_capital.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
+# !/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+"""
+Created on 03 01 2025
+
+@author: Kareiman Altayeb
+"""
+
```
Comment (Frank2446-dotcom) [inline solutions/tests/test_convert_to_capital.py:16]: The test file’s docstring is minimal and could be expanded to better describe the purpose of the tests.

File: solutions/tests/test_convert_to_capital.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+import unittest
+
+# To test convert_to_capital
+from solutions.convert_to_capital import convert_to_capital
+
+
+class TestConvertCapitalLetters(unittest.TestCase):
+    "Tests convert_to_capital function"
```
Comment (Frank2446-dotcom) [inline solutions/tests/test_convert_to_capital.py:24]: Improvement: The test class name TestConvertCapitalLetters is clear, but it could be improved to use a more conventional PascalCase naming style.

Action: Use PascalCase for the test class name, e.g., TestConvertToCapital.

class TestConvertToCapital(unittest.TestCase):

**Review by Frank2446-dotcom — CHANGES_REQUESTED** [review]

File: solutions/tests/test_convert_to_capital.py (hunk 11 above)
Comment (Kareiman) [reply solutions/tests/test_convert_to_capital.py:24]: done

File: solutions/convert_to_capital.py (hunk 8 above)
Comment (Kareiman) [reply solutions/convert_to_capital.py:10]: I added more description but explaining the edge behavior is mentioned within the tests

File: solutions/convert_to_capital.py (hunk 9 above)
Comment (Kareiman) [reply solutions/convert_to_capital.py:38]: I already call strip() when checking for empty strings, so there’s no need to call it again on user_text, it is redundant.

Comment (Kareiman) [reply solutions/convert_to_capital.py:38]: I already call strip() when checking for empty strings, so there’s no need to call it again on user_text., so it is redundant.

File: solutions/convert_to_capital.py (hunk 7 above)
Comment (Kareiman) [reply solutions/convert_to_capital.py:13]: done.

File: solutions/tests/test_convert_to_capital.py (hunk 10 above)
Comment (Kareiman) [reply solutions/tests/test_convert_to_capital.py:16]: The purpose of the edge, classical cases and defensive tests is explained below

**Review by Frank2446-dotcom — APPROVED** [review]
Well done

**Review by Emanfalouji — APPROVED** [review]
Great work kariman ..
